pkg/gateway/model_build_targets.go

```diff
@@ -111,19 +112,38 @@ func (t *latticeTargetsModelBuildTask) buildLatticeTargets(ctx context.Context)
 		return errors.New(errmsg)
 	}
 
-	portAnnotations := undefinedPort
-	serviceExport := &mcs_api.ServiceExport{}
-	err = t.Client.Get(ctx, namespacedName, serviceExport)
-	if err != nil {
-		glog.V(6).Infof("Failed to find Service export in the DS. Name:%v, Namespace:%v - err:%s\n ", t.tgName, t.tgNamespace, err)
-	} else {
-		// TODO: Change the code to support multiple comma separated ports instead of a single port
-		//portsAnnotations := strings.Split(serviceExport.ObjectMeta.Annotations["multicluster.x-k8s.io/Ports"], ",")
-		portAnnotations, err = strconv.ParseInt(serviceExport.ObjectMeta.Annotations[portAnnotationsKey], 10, 64)
+	definedPorts := make([]int64, undefinedPort)
+	if tg.ByServiceExport {
+		serviceExport := &mcs_api.ServiceExport{}
+		err = t.Client.Get(ctx, namespacedName, serviceExport)
 		if err != nil {
-			glog.V(6).Infof("Failed to read Annotaions/Port:%v, err:%s\n ", serviceExport.ObjectMeta.Annotations[portAnnotationsKey], err)
+			glog.V(6).Infof("Failed to find Service export in the DS. Name:%v, Namespace:%v - err:%s\n ", t.tgName, t.tgNamespace, err)
+		} else {
+			portsAnnotations := strings.Split(serviceExport.ObjectMeta.Annotations[portAnnotationsKey], ",")
+
+			for _, portAnnotation := range portsAnnotations {
+				definedPort, err := strconv.ParseInt(portAnnotation, 10, 64)
+				if err != nil {
+					glog.V(6).Infof("Failed to read Annotations/Port:%v, err:%s\n ", serviceExport.ObjectMeta.Annotations[portAnnotationsKey], err)
+				}
+				definedPorts = append(definedPorts, definedPort)
+			}
+			glog.V(6).Infof("Build Targets - portAnnotations: %v \n", definedPorts)
+		}
+	} else if tg.ByBackendRef {
+		if t.route != nil && t.route.Spec().Rules() != nil && 0 < len(t.route.Spec().Rules()) {
+			definedPorts = []int64{}
+			for _, rule := range t.route.Spec().Rules() {
+				for _, ref := range rule.BackendRefs() {
+					if ref.Port() != nil && string(ref.Name()) == t.tgName && string(*ref.Namespace()) == t.tgNamespace {
+						definedPorts = append(definedPorts, int64(*ref.Port()))
+					}
+				}
+			}
 		}
```

Comment:
You could also encapsulate line 134-143 into a helper function to make code more clear:
```
definedPorts = getDefinedPortsFromK8sRouteRules(t.route, t.tgName, t.tgNamespace)
```

```
func getDefinedPortsFromK8sRouteRules(route core.Route, tgName string, tgNamespace string) []int64 {
	if route == nil || route.Spec() == nil || route.Spec().Rules() == nil || len(route.Spec().Rules()) == 0 {
		return []int64{}
	}
	var definedPorts []int64
	for _, rule := range route.Spec().Rules() {
		for _, ref := range rule.BackendRefs() {
			if ref.Port() != nil && string(ref.Name()) == tgName && string(*ref.Namespace()) == tgNamespace {
				definedPorts = append(definedPorts, int64(*ref.Port()))
			}
		}
	}
	return definedPorts
}
``` 

(BTW,  t.tgName is actually k8sService name and t.tgNamespace is actually k8sService's namespace)

line89 `tgName := latticestore.TargetGroupName(t.tgName, t.tgNamespace)` is really confusing...